lib/src/main/java/com/auth0/jwt/impl/PayloadSerializer.java

```diff
@@ -44,20 +45,66 @@ public void serialize(ClaimsHolder holder, JsonGenerator gen, SerializerProvider
                     }
                     break;
                 default:
-                    gen.writeFieldName(e.getKey());
-                    if (e.getValue() instanceof Date) { // true for EXPIRES_AT, ISSUED_AT, NOT_BEFORE
-                        gen.writeNumber(dateToSeconds((Date) e.getValue()));
-                    } else {
-                        gen.writeObject(e.getValue());
-                    }
+                    handleInstantSerialization(e, gen);
                     break;
             }
         }
 
         gen.writeEndObject();
     }
 
-    private long dateToSeconds(Date date) {
-        return date.getTime() / 1000;
+    /**
+     * Serializes {@linkplain Instant} to epoch second values, traversing maps and lists as needed.
+     * @param entry the entry to serialize
+     * @param gen the JsonGenerator to use for JSON serialization
+     * @throws IOException
+     */
+    private void handleInstantSerialization(Map.Entry<String, Object> entry, JsonGenerator gen) throws IOException {
```

Comment:
This does more than just handling "instant serialization". In fact, it's handling all the types as well. I suggest you rename this to something different like "handleSerialization".

Something I just came across is an official module we can register on Jackson's initialization to serialize/deserialize Java8 Date types (and more Java8 classes). Search for `jackson-datatype-jsr310`.

Also, I found a v3 of Jackson is going to support this natively as the min req java version is 8. However, I couldn't find it listed under the "actively developed" branches. Still, worth to check so we can't avoid all this custom serialization code. Besides this file which we will discuss when you're back, you can resolve the comments for the rest of the review.

Reply:
I renamed it to `handleSerialization` as suggested. Regarding the JSR310 module for Jackson, I agree we should investigate and see if we can use it and other OOTB serialization behaviors. I started looking into it but since we are already doing custom serialization, I updated it for now. I think it probably makes sense to evaluate our serialization options separately and make that a separate PR/change if we can.